internal/jsonwall/jsonwall.go

```diff
@@ -0,0 +1,307 @@
+	Version string `json:"jsonwall"`
+	Schema  string `json:"schema,omitempty"`
+	Count   int    `json:"count"`
+}
+
+const jsonwallVersion = "1.0"
+
+func (dbw *DBWriter) writeHeader(w io.Writer, count int) (int, error) {
+	data, err := json.Marshal(&jsonwallHeader{
+		Version: jsonwallVersion,
+		Schema:  dbw.options.Schema,
+		Count:   count,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("internal error: cannot marshal database header")
+	}
+	return w.Write(append(data, '\n'))
+}
+
+// WriteTo assembles the current database state and writes it to w.
+func (dbw *DBWriter) WriteTo(w io.Writer) (n int64, err error) {
+	m, err := dbw.writeHeader(w, len(dbw.entries)+1)
+	n += int64(m)
+	sort.Sort(sortableEntries(dbw.entries))
+	if err != nil {
+		return n, err
+	}
+	for _, entry := range dbw.entries {
+		m, err := w.Write(entry)
+		n += int64(m)
+		if err == nil {
+			m, err = w.Write([]byte{'\n'})
+			n += int64(m)
+		}
+		if err != nil {
+			return n, err
+		}
+	}
+	return n, nil
+}
+
+// ReadDB reads into memory the database from the provided r.
+func ReadDB(r io.Reader) (*DB, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	record := 0
+	for i := range data {
+		if data[i] == '\n' {
+			record = i + 1
+			break
+		}
+	}
+	var header jsonwallHeader
+	err = json.Unmarshal(data[:record], &header)
+	if err != nil {
+		return nil, fmt.Errorf("invalid database content")
+	}
+	if !strings.HasPrefix(header.Version, jsonwallVersion[:strings.Index(jsonwallVersion, ".")+1]) {
+		return nil, fmt.Errorf("unsupported database format: %q", header.Version)
+	}
+	if header.Count > len(data)/8 {
+		// The header helps pre-allocating an index of the right size,
+		// but it could trivially be abused to cause an OOM situation.
+		header.Count = 0
+	}
+	db := &DB{schema: header.Schema, data: data}
+	db.index = make([]int, 0, header.Count)
+	for i := range data {
+		if data[i] == '\n' {
+			db.index = append(db.index, i+1)
+		}
+	}
+	db.count = len(db.index) - 1 // Last item is end of buffer.
+	return db, nil
+}
+
+// DB holds an in-memory read-only database ready for querying.
+type DB struct {
+	schema string
+	data   []byte
+	index  []int
+	count  int
+}
+
+// Schema returns the optional schema value that was provided when writing the database.
+// This information is available to help with the decoding of the data, and is not
+// internally interpreted in any way.
+func (db *DB) Schema() string {
+	return db.schema
+}
+
+func (db *DB) prefix(value any) ([]byte, error) {
+	data, err := json.Marshal(value)
+	if err != nil || len(data) == 0 || data[0] != '{' {
+		return nil, fmt.Errorf("invalid database search value: %#v", value)
+	}
+	data[len(data)-1] = ','
+	return data, nil
+}
+
+var ErrNotFound = fmt.Errorf("value not found in database")
+
+const notFound = -2
+
+func (db *DB) search(prefix []byte) (i int, err error) {
+	i = sort.Search(db.count, func(i int) bool {
+		res := bytes.Compare(db.data[db.index[i]:], prefix) >= 0
```

Comment:
Appreciate the observation, but am I'm not sure I get what's the concern. The slicing operation just copies the original slice header (3 words) and replaces the data pointer and length. The comparison will stop at the shortest string, typically the prefix which is necessarily a single line long due to our own marshaling. Am I missing something?

Reply:
Just discovered te flaw in my argument ... read on if you want to know.
```
l1 + l2 + l3 + ... vs. prefix
l2 + l3 + ... vs. prefix
l3 + ... vs. prefix
```
Which, as you say, in practise is equivalent to:
```
l1 vs. prefix
l2 vs. prefix
l3 vs. prefix
```
However, in theory (the almost impossibly lucky case) if the prefix were to be longer than line its compared against, and happen to match that line, the search would continue into the next line.

For example:

l1 = "abc"
l2 = "def"
prefix = "abcd"

In this scenario, a match will be reported on l1. 

**This is not true because the newline will prevent the match to cross into the next line ...**

doh.

Reply:
Ah, yes indeed.